List comprehension

ninegua opened the pull request:
This should serve as an answer or a partial answer to  #504 #505 #1730 #2094 #2168 #2262 #2301

Several notes:
1. Nesting is supported, but nested list is not, which I believe is an orthogonal issue.
2. *Update*: Python 3 before Python 3.6 is ~~known not to work due to http://bugs.python.org/issue19611~~ now working with ab69d62 
3. Some initial code was based on the old pull request #1045

datnamer commented:
may be related, but do you think there is any hipe for closures and higher order functions or not with current architecture?

ninegua commented:
Do we care about supporting 3.4 and 3.5? Python 3.6 has removed MAKE_CLOSURE opcode, so I've not implemented it.

ninegua commented:
BTW, there is a valid reason that it is difficult to support Python 3 versions before 3.6: http://bugs.python.org/issue19611

ninegua commented:
CI build using Python 2.7 and Python 3.6 have passed, and Python 3.4 and 3.5 are failing. One solution is to add version checks to skip the comprehension unit test when Python 3 version less than 3.6 is being used in order to make it pass. Suggestions?

ehsantn commented:
@ninegua since comprehension has known issues on <3.6 versions, I think it's fine to add version checks and skip them (at least for now).

ninegua commented:
The CI failure seems like a random behavior related to how fusion is performed. @ehsantn @DrTodd13 We need to look at this further.

seibert commented:
Given that Python 3.4/3.5 are the bulk of our Python 3 users, I'd like to see if we can figure out how to make this PR work there as well.

ninegua commented:
@seibert I figured out a way to get around the problem with 3.4/3.5, so it is working now.

seibert commented:
Great!  I've assigned @sklam to review.

sklam commented:
This is awesome.  I have started my review by trying out the new feature and realize that this PR will allow closure:

```python

In [1]: from numba import njit

In [2]: @njit
   ...: def foo(n):
   ...:     def inner(x):
   ...:         return x + n
   ...:     return inner(n)
   ...: 

In [3]: foo(12)
Out[3]: 24
```

We should add test around the closure support.   

The other thing I notice is:

```python

sklam commented:
I can confirm that this problem:

```python

In [6]: @njit
   ...: def foo(n):
   ...:     a = [i for i in range(n)]
   ...:     b = [i for i in range(n)]
   ...:     for i, j in zip(a, b):
   ...:         print(i, j)
   ...:         

In [7]: foo(3)
...
TypingError: Failed at nopython (nopython frontend)
cannot unify zip(iter(list(undefined)), iter(list(undefined))) and zip(iter(list(int64)), iter(list(int64))) for '$53.4', defined at <ipython-input-6-2cf009e5a068> (5)
File "<ipython-input-6-2cf009e5a068>", line 5
[1] During: resolving callee type: Function(<class 'zip'>)
[2] During: typing of call at <ipython-input-6-2cf009e5a068> (5)
```

is independent of the list-comp feature added in this PR.

ninegua commented:
@sklam The closure support provided by this patch is limited in that it relies on inlining to support the full semantics of closure var and free vars. So if a closure isn't immediately called after creation, say it is being returned or passed as argument to another call, the compilation will still fail. So the scope here is only limited to comprehension supported.

The issue you raised looks like a problem with type inference. The IR after closure inline looks something like this:

```
$61 = build_list(items=[])
...
$65 = $61
...
$77 = cast(value=$65)
...
$0.16 = $77 
...
b = $0.16    
```

where `$61` and `$65` are defined in the same block, `$77` and `$0.16` are defined in the same block.

Now if I change the assignment of `$77` to `$77 = cast(value=$61)` or `$77 = $61`, then the inference would work. It doesn't work when I change to `$77 = $65`. 

Also it works if I change the original program to directly index list instead of using `zip`:

```
    for i in range(n):
        print(a[i], b[i])
```

I'm not familiar with the inference part of Numba, so your help would be greatly appreciated!

ninegua commented:
I added a few test cases for closure inline. In the process, I also fixed an issue with assigning to a free variable from within the closure in commit 3c47edd, without which `test_inner_function_with_closure_2` will fail. I don't know how much Numba is relying on the IR staying true to SSA, but this should only affect code that has cell vars. @sklam

ninegua commented on numba/tests/test_comprehension.py:
```diff
@@ -0,0 +1,256 @@
+from __future__ import print_function
```
I think it'll actually make sense to add more variants to this file when other comprehensions are becoming supported, since they are all comprehensions, and they more or less will exercise the same inline closure pass. 

I'm actually adding tests for "array comprehension" (turn `array([...])` into direct array creation without intermediate list) to this file in an upcoming PR :-)

stuartarchibald replied:
This sounds fine noting it will likely be the same code path backing them. I'm looking forward to "array comprehensions" :)

stuartarchibald commented:
@ninegua Thanks for the fix ups in 8a30b3c. Am happy to merge.
